hal/armv7a/zynq7000/_init.S

```diff
@@ -0,0 +1,322 @@
+	orr r1, r1, #(3 << 3)            /* Outer cacheability                            */
+	mcr p15, 0, r1, c2, c0, 0        /* Write Translation Table Base Register 0       */
+	mcr p15, 0, r1, c2, c0, 1        /* Write Translation Table Base Register 1       */
+
+	/* Set all Domains to Client */
+	ldr r1, =0x55555555
+	mcr p15, 0, r1, c3, c0, 0        /* Write Domain Access Control Register */
+
+	/* Enable L1 Caches */
+	mrc p15, 0, r1, c1, c0, 0        /* Read SCTLR (System Control Register) data  */
+	orr r1, r1, #(0x1 << 2)          /* Enable data cache                          */
+	orr r1, r1, #(0x1 << 12)         /* Enable instruction cache                   */
+	orr r1, r1, #(0x1 << 11)         /* Enable branch prediction                   */
+	bic r1, r1, #(0x1 << 28)         /* Disable TEX remap                          */
+	mcr p15, 0, r1, c1, c0, 0        /* Write SCTLR (System Control Register) data */
+
+
+	/* Enable MMU */
+	mrc p15, 0, r1, c1, c0, 0        /* Read Control Register configuration data  */
+	orr r1, r1, #1                   /* Enable MMU settinh bit 0                  */
+	mcr p15, 0, r1, c1, c0, 0        /* Write Control Register configuration data */
+	dsb
+	isb
+
+
+	/* Map perpehrals adresses */
+	/* Map UART0 4 KB P 0xE0000000 -> V CEIL(_end, SIZE_PAGE) */
+	ldr r0, =(VADDR_UART0 - VADDR_KERNEL)
+	lsr r0, #12
+	lsl r0, #2
+	ldr r1, =ADDR_TTL2_K
+	add r0, r0, r1
+	ldr r1, =(PADDR_UART0 | 0x12)
+	str r1, [r0], #4
+
+	/* Map UART1 4KB P 0xE0001000 -> V CEIL(_end + SIZE_PAGE, SIZE_PAGE) */
+	ldr r1, =(PADDR_UART1 | 0x12)
+	str r1, [r0], #4
+
+	/* Map GIC 16 KB after UARTs */
+	mrc p15, 4, r1, c15, c0, 0           /* Get GIC paddr */
+	lsr r1, #16
+	lsl r1, #16
+	orr r1, r1, #0x12
+	mov r2, #(1 << 12)
+	bl _cpy4
+
+	/* Map SLCR after GIC */
+	ldr r1, =0xf8000012
+	str r1, [r0], #4
+
+
+	/* Setup stacks */
+	eor r0, r0        /* initial SP */
+
+	/* FIQ mode stack */
+	msr CPSR_c, #(FIQ_MODE | NO_INT)
+	mov sp, r0
+	sub r0, r0, #0x20
+
+	/* IRQ mode stack */
+	msr CPSR_c, #(IRQ_MODE | NO_INT)
+	mov sp, r0
+	sub r0, r0, #0x100
+
+	/* Supervisor mode stack */
+	msr CPSR_c, #(SVC_MODE | NO_INT)
+	mov sp, r0
+	sub r0, r0, #0x40
+
+	/* Undefined mode stack */
+	msr CPSR_c, #(UND_MODE | NO_INT)
+	mov sp, r0
+	sub r0, r0, #0x40
+
+	/* Abort mode stack */
+	msr CPSR_c, #(ABT_MODE | NO_INT)
+	mov sp, r0
+	sub r0, r0, #0x40
+
+	/* System mode stack */
+	msr CPSR_c, #(SYS_MODE | NO_INT)
+	mov sp, r0
+
+
+	/* Enable FPU */
+	mrc p15, 0, r0, c1, c0, 2                 /* Read CPACR into R0                                 */
+	orr r0, r0, #((0x3 << 22) | (0x3 << 20))  /* Set CP11 and CP10: Privileged and User mode access */
+	mcr p15, 0, r0, c1, c0, 2                 /* Write R0 to CPACR                                  */
+	vmrs r0, fpexc
+	orr r0, r0, #(0x1 << 30)                  /* FPU enable bit                                     */
+	vmsr fpexc, r0
+
+	ldr pc, =main
+
+#include "../_interrupts.S"
+#include "../_hal.S"
```

Comment:
We can include this files by compiling them and linking together, it's somewhat cleaner solution. Especially that _interrupts.S and _hal.S are common files

Reply:
I agree with you.
However, interrupts need some change to be done. For example there is direct call to GIC controller in `_interrupts_dispatch` https://github.com/phoenix-rtos/phoenix-rtos-kernel/blob/eaffcef02fcd21966b5be870d7c4e371b9a7bfd4/hal/armv7a/imx6ull/_init.S#L859-L877
It should be moved to interrupt controller in imx6ull and Zynq-7000. 
When it will be done, I will add them to compilation process and remove `include`